### PR TITLE
fix: apply pre-conditions first on object metadata

### LINKED ISSUE #12543: GetObject returns different response to Amazon S3
When we do a GetObject call against MinIO with both an invalid etag in the If-Match field, and a range that's past the end of the object data in the Range field, we get a different error from MinIO than from Amazon S3.

## Expected Behavior
Amazon S3 returns the following error:
An error occurred (PreconditionFailed) when calling the GetObject operation: At least one of the pre-conditions you specified did not hold

## Current Behavior
MinIO returns:
An error occurred (InvalidRange) when calling the GetObject operation: The requested range is not satisfiable

## Steps to Reproduce (for bugs)
Create an object with data using put_object, then run a get_object request setting both an invalid etag and range past the end of the object in the request_headers field.

Code sample to repro:

```py
import io

from minio import Minio
from minio.error import S3Error

def main():
    # Create a client with the MinIO server playground, its access key
    # and secret key.
    client = Minio(
        "play.min.io",
        access_key="Q3AM3UQ867SPQQA43P2F",
        secret_key="zuf+tfteSlswRu7BJ86wekitnifILbZam1KYY3TG",
    )

    # Upload empty folder
    result = client.put_object(
        "test", "object", io.BytesIO(b"data"), 4,
    )
    print("created {0} object; etag: {1}, version-id: {2}".format(
            result.object_name, result.etag, result.version_id,
        ),
    )

    client.get_object("test", "object", request_headers={ 'If-Match': 'invalid_etag', 'Range': 'bytes=4-10' })

if __name__ == "__main__":
    try:
        main()
    except S3Error as exc:
        print("error occurred.", exc)
```

## Your Environment

* Version used (`minio --version`):
minio version RELEASE.2020-06-22T03-12-50Z

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
fix: apply pre-conditions first on object metadata

## Motivation and Context
This change in error flow complies with AWS S3 behavior
for applications depending on specific error conditions.

fixes #12543

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — COMMENTED** [review]
This is also an optimization such that we don't have to invoke a Reader() stream before checking preconditions. Allowing for a more expected optimal behavior where metadata is verified first for expected conditions before proceeding to read the stream.

**Review by poornas — APPROVED** [review]
LGTM

**Comment by klauspost** [comment]
So this is a regression from #12487

I am probably missing something @harshavardhana, but why isn't this reintroducing the same problem that #12487 fixes (and is #12485 the correct fix)?

**Comment by harshavardhana** [comment]
> So this is a regression from #12487
> 
> I am probably missing something @harshavardhana, but why isn't this reintroducing the same problem that #12487 fixes (and is #12485 the correct fix)?

This is not a regression or anything we have had this behavior for long. We simply didn't do it because of a previous implementation baggage that doesn't exist for some time now.

We simply used to validate ranges before preconditions. We could have validated preconditions before validating ranges.

This has nothing to do with pool expansion and duplicate objects @klauspost - this fix is simply a top level fix on error flow on what should be considered as an error first.

Check the original issue #12543

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

File: cmd/erasure-server-pool.go [hunk 1]
```diff
@@ -748,24 +678,50 @@ func (z *erasureServerPools) GetObjectInfo(ctx context.Context, bucket, object s
...
+		return objInfo, err
+	}
+
+	object = encodeDirObject(object)
+
+	if z.SinglePool() {
+		return z.serverPools[0].GetObjectInfo(ctx, bucket, object, opts)
+	}
+
+	opts.NoLock = true // avoid taking locks at lower levels for multi-pool setups.
```
Comment (klauspost) [inline cmd/erasure-server-pool.go]: Check if it is already true, so parent has locked it.

Comment (harshavardhana) [reply cmd/erasure-server-pool.go]: @klauspost okay

**Review by klauspost — APPROVED** [review]
LGTM

**Comment by harshavardhana** [comment]
Going to add a new mint-test for this in minio-py

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Pull request merged**